### ISSUE TITLE
🧪 Performance Benchmarks: Neural network operations

### ISSUE THREAD

**Issue by mikkihugo** [mikkihugo/claude-code-zen#234]
## 📋 Summary

Implement comprehensive performance benchmarks for neural network operations to establish baselines, detect regressions, and optimize critical paths.

### Benchmark Categories:
- [ ] **Model Operations**
  - Forward pass latency
  - Backward pass performance
  - Batch processing throughput

- [ ] **Training Performance**
  - Epochs per second
  - Gradient computation time
  - Parameter update speed

- [ ] **Data Processing**
  - Dataset loading time
  - Preprocessing overhead
  - Feature engineering cost

- [ ] **Memory Usage**
  - Model memory footprint
  - Training memory growth
  - Peak memory usage

### 1. Model Benchmarks
```rust
use criterion::{black_box, criterion_group, criterion_main, Criterion};

fn benchmark_mlp_forward(c: &mut Criterion) {
    c.bench_function("mlp_forward_1000", |b| {
        let model = create_mlp(1000, 100, 10);
        let input = random_input(1000);
        b.iter(|| model.forward(black_box(&input)));
    });
}
```

### 2. Training Benchmarks
- Single epoch timing
- Full training runs
- Different batch sizes
- Various optimizers

### 3. Scaling Benchmarks
- Input size scaling
- Model size scaling
- Batch size scaling
- Thread count scaling

### 4. Memory Benchmarks
- Allocation patterns
- Memory fragmentation
- Cache efficiency
- SIMD utilization

### 5. Comparison Benchmarks
- ruv-FANN baseline
- Different model architectures
- Configuration variations

## 🛠️ Technical Setup

```toml
[dev-dependencies]
criterion = { version = "0.5", features = ["html_reports"] }
divan = "0.1"
pprof = { version = "0.14", features = ["flamegraph"] }
```

### Benchmark Organization
```
benches/
├── model_benchmarks.rs
├── training_benchmarks.rs
├── data_benchmarks.rs
└── memory_benchmarks.rs
```

## 📊 Success Criteria
- [ ] Baseline performance established
- [ ] Regression detection in CI
- [ ] Performance reports generated
- [ ] Optimization opportunities identified

## 🔗 Parent Issue
Part of #229
